applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp

```diff
@@ -812,6 +812,80 @@ Matrix ConstitutiveLawUtilities<TVoigtSize>::CalculateLinearPlasticDeformationGr
 /***********************************************************************************/
 /***********************************************************************************/
 
+template<SizeType TVoigtSize>
+Matrix ConstitutiveLawUtilities<TVoigtSize>::CalculateExponentialElasticDeformationGradient(
+    const MatrixType& rElasticTrial,
+    const BoundedVectorType& rPlasticPotentialDerivative,
+    const double PlasticConsistencyFactorIncrement,
+    const MatrixType& rRe
+    )
+{
+    // Define elastic deformation gradient
+    MatrixType elastic_deformation_gradient(Dimension, Dimension);
+
+    // Define plastic flow
+    const BoundedMatrixType plastic_flow = PlasticConsistencyFactorIncrement * MathUtils<double>::StrainVectorToTensor<BoundedVectorType, MatrixType>(rPlasticPotentialDerivative);
+
+    // Declare the different eigen decomposition matrices
+    BoundedMatrixType eigen_values_matrix, eigen_vectors_matrix;
+
+    // We compute the exponential matrix
+    // Decompose matrix
+    MathUtils<double>::GaussSeidelEigenSystem(plastic_flow, eigen_vectors_matrix, eigen_values_matrix, 1.0e-16, 20);
+
+    // Calculate the eigenvalues of the E matrix
+    for (std::size_t i = 0; i < Dimension; ++i) {
+        eigen_values_matrix(i, i) = std::exp(-eigen_values_matrix(i, i));
+    }
+
+    // Calculate exponential matrix
+    MathUtils<double>::BDBtProductOperation(elastic_deformation_gradient, eigen_values_matrix, eigen_vectors_matrix);
+
+    // Pre and post multiply by Re
+    elastic_deformation_gradient = prod(elastic_deformation_gradient, rRe);
+    elastic_deformation_gradient = prod(trans(rRe), elastic_deformation_gradient);
+    elastic_deformation_gradient = prod(rElasticTrial, elastic_deformation_gradient);
+
+    return elastic_deformation_gradient;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TVoigtSize>
+Matrix ConstitutiveLawUtilities<TVoigtSize>::CalculateDirectElasticDeformationGradient(
+    const MatrixType& rElasticTrial,
```

Comment:
we could add a ref to the book pg?

Reply:
This formula in fact is not in the book, but can be deduced from the book:

![imagen](https://user-images.githubusercontent.com/19991680/64486161-cc9f6100-d229-11e9-8917-692c36519937.png)